File: ClipperApp/NotOverlapTest.cpp
//
// Created by Nail Sharipov on 29.11.2023.
//

#include "NotOverlapTest.h"
#include "Util.h"
#include "include/clipper2/clipper.h"
#include <iostream>
#include <chrono>
#include <iomanip>

using namespace Clipper2Lib;

// A grid of not overlapping squares.
void NotOverlapTest::run(int n) {
    std::cout << "Start NotOverlap Test\n";

    auto subj = manySquares({0, 0}, 20, 30, n);

    int it_count = (int)(1000.0 / (double)n);
    it_count = it_count * it_count;
    it_count = it_count == 0 ? 1 : it_count;
    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < it_count; ++i) {
        Paths64 solution = Union(subj, FillRule::NonZero);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    double time = elapsed.count() / (double)it_count;

    int count = n * n;
    double count_log = log2(count);

    double time_log = log10(time);

    std::cout << n << "(" << std::fixed << std::setprecision(1) << count_log << ")"
              << "     - " <<std::fixed << std::setprecision(6)<< time << "(" << std::fixed << std::setprecision(1) << time_log << ")\n";
}

File: iOverlayAppRust/src/test/not_overlap_test.rs
use std::time::Instant;
use i_float::point::IntPoint;
use i_overlay::core::fill_rule::FillRule;
use i_overlay::core::overlay::{Overlay, ShapeType};
use i_overlay::core::overlay_rule::OverlayRule;
use crate::test::util::Util;

pub(crate) struct NotOverlapTest;

// 2
// Union:

//   3(3.2)        - 0.000009(-5.0)
//   6(5.2)        - 0.000035(-4.5)
//  12(7.2)        - 0.000165(-3.8)
//  25(9.3)        - 0.000845(-3.1)
//  50(11.3)       - 0.003907(-2.4)
//  100(13.3)      - 0.018785(-1.7)
//  200(15.3)      - 0.087480(-1.1)
//  400(17.3)      - 0.392822(-0.4)
//  800(19.3)      - 1.697748(0.2)
// 1600(21.3)      - 7.326802(0.9)
// 3200(23.3)      - 29.108937(1.5)
// 6400(25.3)     - 130.862234(2.1)

// A grid of not overlapping squares.
impl NotOverlapTest {
    pub(crate) fn run(n: usize) {
        println!("Start NotOverlap Test");
        let subj_paths = Util::many_squares(IntPoint::new(0, 0), 20, 30, n);

        let start = Instant::now();

        let it_count = ((1000.0 / (n as f64)) as usize).max(1);
        let it_count= it_count * it_count;
        for _ in 0..it_count {
            let mut overlay = Overlay::new(4 * n * n);
            overlay.add_paths(&subj_paths, ShapeType::Subject);

            let graph = overlay.build_graph(FillRule::NonZero);
            _ = graph.extract_shapes(OverlayRule::Subject);
        }

        let duration = start.elapsed();
        let time = duration.as_secs_f64() / it_count as f64;

        let count = n * n;
        let count_log = (count as f64).log2();
        let time_log = time.log10();

        println!("{}({:.1})     - {:.6}({:.1})", n, count_log, time, time_log);
    }
}